velox/substrait/SubstraitToVeloxExpr.cpp

```diff
@@ -167,19 +178,25 @@ SubstraitVeloxExprConverter::toVeloxExpr(
     case ::substrait::Expression::FieldReference::ReferenceTypeCase::
         kDirectReference: {
       const auto& directRef = substraitField.direct_reference();
-      int32_t colIdx = substraitParser_.parseReferenceSegment(directRef);
-      const auto& inputNames = inputType->names();
-      const int64_t inputSize = inputNames.size();
-      if (colIdx <= inputSize) {
-        const auto& inputTypes = inputType->children();
-        // Convert type to row.
-        return std::make_shared<core::FieldAccessTypedExpr>(
-            inputTypes[colIdx],
-            std::make_shared<core::InputTypedExpr>(inputTypes[colIdx]),
-            inputNames[colIdx]);
-      } else {
-        VELOX_FAIL("Missing the column with id '{}' .", colIdx);
+      core::FieldAccessTypedExprPtr fieldAccess{nullptr};
+      const auto* tmp = &directRef.struct_field();
+
+      auto inputColumnType = inputType;
+      for (;;) {
+        auto idx = tmp->field();
+        fieldAccess = makeFieldAccessExpr(
+            inputColumnType->nameOf(idx),
+            inputColumnType->childAt(idx),
+            fieldAccess);
+
+        if (!tmp->has_child()) {
+          break;
+        }
+
+        inputColumnType = asRowType(inputColumnType->childAt(idx));
+        tmp = &tmp->child().struct_field();
```

Comment:
Is tmp->child() guaranteed to be a direct reference? Where can I read about direct_reference?

Reply:
1. struct_field() will return class `StructField`.  [StructField link](https://github.com/facebookincubator/velox/blob/main/velox/substrait/proto/substrait/algebra.proto#L970)
2. StructField child() will return `ReferenceSegment` [struct_field.child() link](https://github.com/facebookincubator/velox/blob/main/velox/substrait/proto/substrait/algebra.proto#L987)
3. ReferenceSegment will represent one type of  `MapKey` `StructField` `ListElement`. [ReferenceSegment define](https://github.com/facebookincubator/velox/blob/main/velox/substrait/proto/substrait/algebra.proto#L967)
4. direct_reference is also ReferenceSegment. [direct_reference link](https://github.com/facebookincubator/velox/blob/main/velox/substrait/proto/substrait/algebra.proto#L1074)